lib/battle/BattleHex.cpp

```diff
@@ -12,6 +12,9 @@
 
 VCMI_LIB_NAMESPACE_BEGIN
 
+//Constexpr defined array with all directions used in battle
+static constexpr BattleHex::EDir battleDirections[] = {BattleHex::TOP_LEFT, BattleHex::TOP_RIGHT, BattleHex::RIGHT, BattleHex::BOTTOM_RIGHT, BattleHex::BOTTOM_LEFT, BattleHex::LEFT};

```

Comment:
@IvanSavenko will rename. It is not a global-level, it is unit-level constexpr definition. As class member or static method it will be promoted to other files, and can lead to unsuccessful linking (I actually tried as private static constexpr member firstly, and failed a linking next. I think as static constexpr unit-only definition it is OK.

I will show you errors later.

Reply:
In general - constexpr members can lead to unsuccessful compilation, and it is better to make it private to only compilation unit (hence, static variable in .cpp file, for example).

Reply:
> As class member or static method it will be promoted to other files, and can lead to unsuccessful linking

Static *member* probably should be *declared* in header but *defined* in header. AFAIK needed even for constexpr, with c++17 (or was it 20?) it is finally possible to mark such member is 'inline' and define in-place

Static *method* should not have these issues and should work either way

Reply:
If something is declared constexpr, it should defined AND declared in one place, AFAIK (at least for c++14). Yes, we can do constexpr method, but what is more optimized?

Reply:
Probably method then? Constexpr or not. Works on any system and such form already in use in our code

Reply:
It should be constexpr to save us from memory burden. I can try later today.

Reply:
done as constexpr method, but I do not know if it is optimized out or not.